specs/AddHostObjectToScriptWithOrigins.md

```diff
@@ -0,0 +1,161 @@
+        Callback<ICoreWebView2FrameEventHandler>(
+            [this](
+                ICoreWebView2* sender,
+                ICoreWebView2FrameEventArgs* args) -> HRESULT
+    {
+        wil::com_ptr<ICoreWebView2Frame> webviewFrame;
+        CHECK_FAILURE(args->get_Frame(&webviewFrame));
+
+        // Wrap host object
+        VARIANT remoteObjectAsVariant = {};
+        m_hostObject.query_to<IDispatch>(&remoteObjectAsVariant.pdispVal);
+        remoteObjectAsVariant.vt = VT_DISPATCH;
+
+        // Create list of origins which will be checked.
+        // iframe will have access to host object only if its origin belongs
+        // to this list.
+        LONG i = 0;
+        SAFEARRAY* list = SafeArrayCreateVector(VT_BSTR, 0, 1);
+        BSTR origin = SysAllocString(L"https://appassets.example");
+        SafeArrayPutElement(list, &i, origin);
+
+        VARIANT originList = {};
+        originList.parray = list;
+        originList.vt = VT_ARRAY;
+
+        // Add host object to a script for iframe access
+        CHECK_FAILURE(
+            webviewFrame->AddHostObjectToScriptWithOrigins(
+                L"sample", &remoteObjectAsVariant, &originList));
+
+        remoteObjectAsVariant.pdispVal->Release();
+        SysFreeString(origin);
+        SafeArrayDestroy(list);
+
+        return S_OK;
+    }).Get(), &m_frameCreatedToken));
+}
+```
+
+
+# Remarks
+List of origins in AddHostObjectToScript will be treated as following:
+ - empty list - no host object will be added for the iframe;
+ - list with origins - during access to host object from iframe the origin will be
+ checked that it belongs to this list;
+ - list with "*" element - host object will be available for iframe for all origins.
+ We suggest not to use this feature without understanding security implications of
+ giving access to host object from from iframes with unknown origins.
+
+
+# API Notes
+See [API Details](#api-details) section below for API reference.
+
+# API Details
+## IDL
+```c#
+/// WebView2Frame provides direct access to the iframes information and use.
+[uuid(18ae830b-8c83-459f-8a8c-95a1022af774), object, pointer_default(unique)]
+interface ICoreWebView2Frame : IUnknown {
+  /// The name of the iframe from the iframe html tag declaring it.
+  [propget] HRESULT Name([ out, retval ] LPWSTR * name);
+  /// Unique frame identifier.
+  [propget] HRESULT Id([ out, retval ] INT32* frameId);
+  /// Add the provided host object to script running in the WebView2 with the
+  /// specified name for the list of the specified origins. The host object
+  /// will be accessible for this iframe only if the iframe's origin during access
+  /// matches one of the origins which are passed.
+  /// \snippet ScenarioAddHostObject.cpp AddHostObjectToScriptWithOrigins
+  /// For more information about host objects navigate to [AddHostObjectToScript]
+  HRESULT AddHostObjectToScriptWithOrigins(
+      [in] LPCWSTR name,
+      [in] VARIANT *object,
+      [in] VARIANT *origins);
```

Comment:
Ok, I do not like how long is code for building VARIANT *origins and agree that it could be done better. But supporting one origin instead of list could introduce ambiguity. For example user makes 2 calls: `AddHostObjectToScriptWithOrigins("name1", object1,  "webview2.com")` and `AddHostObjectToScriptWithOrigins("name1", object2,  "microsoft.com")`  - names are the same, origins different and objects are different. The result of multiple calls in this case should be different then when objects are the same. With origins as a list, we just substitute the object - so the call is kind of atomic and less error prone, also this behavior would conform to current AddHostObjectToScript in WebView2.

Internally I'm converting VARIANT *origins to SAFEARRAY. Is there a way to pass SAFEARRAY from here in theory? Are there any other way to pass a list of strings which is more reliable?

Reply:
Ok, so two options:
1. 
```
HRESULT AddHostObjectToScriptWithOrigin(
      [in] LPCWSTR name,
      [in] VARIANT *object,
      [in] LPCWSTR origin);
```
Replace parameter VARIANT * origins with LPCWSTR origin(renaming method to  AddHostObjectToScriptWithOrigin - without ending s). If user is calling  AddHostObjectToScriptWithOrigin with an object name which already exists  - fail that call always(or fail only if object is different? - is it possible to detect that object is the same?). If user wants to have another origin for existing object it has to be removed or "*" should be used initially as origin. In future we can extend this to support regex  as origin(or I can do it now).

2. 
```
HRESULT AddHostObjectToScriptWithOptions(
      [in] LPCWSTR name,
      [in] VARIANT *object,
      [in] HostObjectOptions options);
```
Where HostObjectOptions ( ICoreWebView2HostObjectOptions? ) is new type with origins list. Which interface to search to see how [in] list was done previously as example?

@david-risney Please let me know which option do you prefer or if there are more of them.

Reply:
I'm hesitant over SAFEARRAY because it uses a SHORT to describe its length and isn't type safe.

A third option could be to use two parameters, one a count and one a LPCWSTR* and we use CoMemAlloc.

I'll send a mail to API review board folks if they have issues, recommendations on this sort of thing.

Reply:
Option 3 was preferred: Using two parameters, one is a count and one is a LPCWSTR*